File: mc/buf.php
<?php
/*
 * A string buffer with functions for parsing
 */
class buf
{
	private $str;
	private $pos;
	private $len;

	/*
	 * Line and column counters
	 */
	private $line = 1;
	private $col = 1;
	private $linelengths = array();

	function __construct($str)
	{
		$this->str = $str;
		$this->pos = 0;
		$this->len = strlen($str);
	}

	function ended()
	{
		return $this->pos >= $this->len;
	}

	function pos()
	{
		return "$this->line:$this->col";
	}

	function peek()
	{
		if($this->pos >= $this->len) {
			return null;
		}
		return $this->str[$this->pos];
	}

	function get()
	{
		if($this->pos >= $this->len) {
			return null;
		}
		$ch = $this->str[$this->pos++];
		if($ch == "\n") {
			$this->line++;
			$this->linelengths[] = $this->col;
			$this->col = 1;
		}
		else {
			$this->col++;
		}
		return $ch;
	}

	function unget($ch)
	{
		$this->pos--;
		assert($this->str[$this->pos] == $ch);
		if($ch == "\n") {
			$this->line--;
			$this->col = array_pop($this->linelengths);
		}
		else {
			$this->col--;
		}
	}

	function context($n = 10)
	{
		$s = '';

		$len = $n;
		$p = $this->pos - $len;
		if($p < 0) {
			$len += $p;
			$p = 0;
		}
		if($len > 0) {
			$s .= substr($this->str, $p, $len);
		}

		$s .= $this->fcontext($n);
		return $s;
	}

	function fcontext($n = 10)
	{
		$s = '{'.$this->str[$this->pos].'}';
		$s .= substr($this->str, $this->pos + 1, $n);
		$s = str_replace(array("\r", "\n", "\t"),
			array("\\r", "\\n", "\\t"), $s);
		return $s;
	}

	/*
	 * Skips any sequence of characters from the given string.
	 * Returns the skipped string.
	 */
	function read_set( $set )
	{
		$s = '';
		while( ($ch = $this->get()) !== null )
		{
			if( strpos( $set, $ch ) === false ) {
				$this->unget( $ch );
				break;
			}
			$s .= $ch;
		}
		return $s;
	}

	function skip_literal($s)
	{
		$n = strlen($s);
		if(!$this->literal_follows($s)) {
			return false;
		}
		$this->pos += $n;
		return true;
	}

	function until_literal( $str )
	{
		$s = '';
		while( !$this->ended() && !$this->literal_follows( $str ) ) {
			$s .= $this->get();
		}
		return $s;
	}

	private function literal_follows($s)
	{
		$n = strlen($s);
		return substr($this->str, $this->pos, $n) == $s;
	}

	function skip_until($ch)
	{
		// Only single characters are supported
		assert(strlen($ch) == 1);

		$s = '';
		while( !$this->ended() && $this->peek() != $ch ) {
			$s .= $this->get();
		}
		return $s;
	}
}

?>

File: src/buf.rs
// String buffer with functions for parsing.
pub struct Buf {
    s: Vec<char>,
    pos: usize,

    line: usize,
    col: usize,
    linelengths: Vec<usize>,
}

#[derive(Debug, Clone)]
pub struct Pos {
    pub line: usize,
    pub col: usize,
}

impl Pos {
    pub fn fmt(&self) -> String {
        format!("{}:{}", self.line, self.col)
    }
}

pub fn new(s: String) -> Buf {
    Buf {
        s: s.chars().collect(),
        pos: 0,

        line: 1,
        col: 1,
        linelengths: Vec::new(),
    }
}

impl Buf {
    pub fn ended(&self) -> bool {
        return !self.more();
    }

    pub fn more(&self) -> bool {
        self.pos < self.s.len()
    }

    pub fn pos(&self) -> Pos {
        Pos {
            line: self.line,
            col: self.col,
        }
    }

    pub fn peek(&self) -> Option<char> {
        if self.more() {
            return Some(self.at(self.pos));
        } else {
            return None;
        }
    }

    fn at(&self, p: usize) -> char {
        self.s[p]
    }

    pub fn get(&mut self) -> Option<char> {
        if !self.more() {
            return None;
        }

        let c = self.at(self.pos);
        self.pos += 1;
        if c == '\n' {
            self.linelengths.push(self.col);
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        return Some(c);
    }

    // pub fn unget(&mut self, c: char) {
    //     assert!(self.pos > 0);
    //     assert!(self.at(self.pos - 1) == c);
    //     self.pos -= 1;
    //     if c as char == '\n' {
    //         self.line -= 1;
    //         self.col = self.linelengths.pop().unwrap();
    //     } else {
    //         self.col -= 1;
    //     }
    // }

    // Skips any sequence of characters from the given string.
    // Returns the skipped string.
    pub fn read_set(&mut self, set: String) -> String {
        let mut s = String::new();
        loop {
            if !self.more() {
                break;
            }
            let next = self.peek().unwrap_or(0 as char);
            if !set.chars().position(|c| c == next).is_some() {
                break;
            }
            s.push(self.get().unwrap());
        }
        return s;
    }

    pub fn skip_literal(&mut self, literal: &str) -> bool {
        if !self.literal_follows(literal) {
            return false;
        }
        for _ in 0..literal.len() {
            self.get();
        }
        return true;
    }

    pub fn until_literal(&mut self, literal: &str) -> String {
        let mut s = String::new();
        while self.more() && !self.literal_follows(literal) {
            s.push(self.get().unwrap());
        }
        return s;
    }

    pub fn literal_follows(&self, s: &str) -> bool {
        if s.len() > self.s.len() - self.pos {
            return false;
        }
        let mut i = 0;
        for c in s.bytes() {
            if self.at(self.pos + i) != c as char {
                return false;
            }
            i += 1;
        }
        return true;
    }

    // Skips characters until the given character is encountered.
    // The given character is left unread.
    // Returns the skipped characters.
    pub fn skip_until(&mut self, ch: char) -> String {
        let mut s = String::new();
        while self.more() && self.peek().unwrap() != ch {
            s.push(self.get().unwrap());
        }
        return s;
    }

    // pub fn context(&self) -> String {
    //     let s = String::new();

    //     let start = if self.pos < 10 { 0 } else { self.pos - 10 };
    //     let end = if self.pos + 10 > self.s.len() {
    //         self.s.len()
    //     } else {
    //         self.pos + 10
    //     };

    //     format!(
    //         "{}[{}]{}",
    //         self.s[start..self.pos].iter().collect::<String>(),
    //         self.s[self.pos],
    //         self.s[self.pos..end].iter().collect::<String>()
    //     );
    //     return s;
    // }

    // pub fn fcontext(self) -> String {
    //     let n = 10;
    //     let s = String::from(format!("[{}]", self.s[self.pos]));
    //     s += &s[self.pos + 1..(n - self.pos - 1)];
    //     return s;
    //     // s = str_replace(array("\r", "\n", "\t"), array(
    //     // 	"\\r",
    //     // 	"\\n",
    //     // 	"\\t"
    //     // ), $s);
    //     // return $s;
    // }
}

#[cfg(test)]
mod tests {
    use super::new;
    #[test]
    fn ended() {
        let mut buf = new(String::from("1"));
        assert_eq!(buf.ended(), false);
        buf.get();
        assert_eq!(buf.ended(), true);
    }

    #[test]
    fn pos() {
        let mut buf = new(String::from("123\n456"));
        let seq = ["1:1", "1:2", "1:3", "1:4", "2:1", "2:2", "2:3"];
        for pos in &seq {
            assert_eq!(buf.pos().fmt(), pos.to_string());
            buf.get();
        }
    }
    #[test]
    fn peek() {
        let mut buf = new(String::from('1'));
        assert_eq!('1', buf.peek().unwrap());
        assert_eq!('1', buf.peek().unwrap());
        buf.get();
        assert_eq!(buf.peek().is_none(), true);
    }
    #[test]
    fn get() {
        let mut buf = new(String::from("123"));
        assert_eq!(buf.get().unwrap(), '1');
        assert_eq!(buf.get().unwrap(), '2');
        assert_eq!(buf.get().unwrap(), '3');
        assert_eq!(buf.get().is_none(), true);
    }
    // #[test]
    // fn unget() {
    //     let mut buf = new(String::from("123"));
    //     assert_eq!('1', buf.get().unwrap());
    //     assert_eq!('2', buf.get().unwrap());
    //     buf.unget('2');
    //     assert_eq!('2', buf.get().unwrap());
    //     assert_eq!('3', buf.get().unwrap());
    //     assert_eq!(true, buf.get().is_none());
    // }
    #[test]
    fn read_set() {
        let mut buf = new(String::from("aba123"));
        assert_eq!("", buf.read_set(String::from("1234567890")));
        assert_eq!("aba", buf.read_set(String::from("abcdef")));
        assert_eq!('1', buf.get().unwrap());
    }
    #[test]
    fn skip_literal() {
        let mut buf = new(String::from("abc123"));
        assert_eq!(false, buf.skip_literal("123"));
        assert_eq!(true, buf.skip_literal("abc"));
        assert_eq!('1', buf.get().unwrap());
    }
    #[test]
    fn until_literal() {
        let mut buf = new(String::from("abc123"));
        assert_eq!("abc", buf.until_literal("123"));
        assert_eq!('1', buf.get().unwrap());
    }
    #[test]
    fn skip_until() {
        let mut buf = new(String::from("abc123"));
        assert_eq!("abc", buf.skip_until('1'));
        assert_eq!("123", buf.skip_until('z'));
    }
    #[test]
    fn literal_follows() {
        let buf = new(String::from("abc123"));
        assert_eq!(true, buf.literal_follows("abc"));
        assert_eq!(false, buf.literal_follows("123"));
    }
}
